src/sagas/task.saga.js

```diff
@@ -0,0 +1,339 @@
+// @flow
+import { select, call, put, take, takeEvery } from 'redux-saga/effects';
+import { buffers, eventChannel, END } from 'redux-saga';
+import { ipcRenderer } from 'electron';
+import * as childProcess from 'child_process';
+import * as path from 'path';
+import chalkRaw from 'chalk';
+
+import {
+  RUN_TASK,
+  ABORT_TASK,
+  COMPLETE_TASK,
+  LAUNCH_DEV_SERVER,
+  completeTask,
+  attachTaskMetadata,
+  receiveDataFromTaskExecution,
+  loadDependencyInfoFromDisk,
+} from '../actions';
+import { getProjectById } from '../reducers/projects.reducer';
+import { getPathForProjectId } from '../reducers/paths.reducer';
+import { isDevServerTask } from '../reducers/tasks.reducer';
+import findAvailablePort from '../services/find-available-port.service';
+import killProcessId from '../services/kill-process-id.service';
+import { isWin, PACKAGE_MANAGER_CMD } from '../services/platform.service';
+
+import type { Task, ProjectType } from '../types';
+import type { Saga } from 'redux-saga';
+
+const chalk = new chalkRaw.constructor({ level: 3 });
+
+export function* launchDevServer({ task }: { task: Task }): Saga<void> {
+  const project = yield select(getProjectById, task.projectId);
+  const projectPath = yield select(getPathForProjectId, task.projectId);
+
+  try {
+    const port = yield call(findAvailablePort);
+    const { args, env } = yield call(
+      getDevServerCommand,
+      task,
+      project.type,
+      port
+    );
+
+    const child = yield call(
+      [childProcess, childProcess.spawn],
+      PACKAGE_MANAGER_CMD,
+      args,
+      {
+        cwd: projectPath,
+        env: { ...getBaseProjectEnvironment(projectPath), ...env },
+      }
+    );
+
+    // Now that we have a port/processId for the server, attach it to
+    // the task. The port is used for opening the app, the pid is used
+    // to kill the process
+    yield put(attachTaskMetadata(task, child.pid, port));
+
+    yield call([ipcRenderer, ipcRenderer.send], 'addProcessId', child.pid);
+
+    const stdioChannel = createStdioChannel(child, {
```

Comment:
Interesting! Never worked with channels in Redux Saga, seems useful though!

Reply:
Hm, looking at [the docs](https://github.com/redux-saga/redux-saga/blob/master/docs/advanced/Channels.md#using-the-eventchannel-factory-to-connect-to-external-events), it seems like we're supposed to `yield call` the channel-creator:

```js
const stdioChannel = yield call(createStdioChannel, child, { ... });
```

Because `call` will work synchronously when it isn't calling a generator or a function that returns a promise, it seems like it works the same way, except I think it returns an effect description instead of the actual channel.

I tried making this change locally, and it still works the same, although it causes 3 tests to fail 😬

Asking for my own education (I'm still reacquainting myself with redux saga), not because I'm requesting any changes.

Reply:
Yeah, makes sense. I've written some saga tests since writing that comment and it's clearer to me now. I think this is good :+1: